File: ts/src/bws/packer/Wrap.ts
/// <reference path="API.ts" />

namespace bws.packer
{
	/**
	 * <p> Wrap represents an act wrap(ping). </p>
	 *
	 * <p> {@link Wrap} is a class represents an act wrapping an {@link Instance} to an {@link Wrapper}.
	 * To represent the relationship, Wrap uses Bridge and Capsular patterns to links and intermediates 
	 * relationship between Wrapper and Instance. </p>
	 *
	 * <p> Wrap also helps packing optimization and 3d-visualization with its own members 
	 * {@link orientation} and position variables {@link x}, {@link y} and {@link z}. </p> 
	 *
	 * @author Jeongho Nam <http://samchon.org>
	 */
	export class Wrap
		extends protocol.Entity
	{
		/**
		 * A wrapper wrapping the {@link instance}.
		 */
		protected wrapper: Wrapper;
		
		/**
		 * An instance wrapped into the {@link wrapper}.
		 */
		protected instance: Instance;

		/**
		 * Coordinate-X of the instance placement in the wrapper.
		 */
		protected x: number;
		
		/**
		 * Coordinate-Y of the instance placement in the wrapper.
		 */
		protected y: number;
		
		/**
		 * Coordinate-Z of the instance placement in the wrapper.
		 */
		protected z: number;

		/**
		 * Placement orientation of wrapped {@link instance}.
		 */
		protected orientation: number;

		/* ============================================================
			CONSTRUCTORS & SETTERS
		===============================================================
			CONSTRUCTORS
		------------------------------------------------------------ */
		/**
		 * Construct from a Wrapper.
		 * 
		 * @param wrapper A wrapper who will contain an instance.
		 */
		public constructor(wrapper: Wrapper);

		/**
		 * Construct from a Wrapper and Instance with its position and default orientation 1. 
		 *
		 * @param wrapper A wrapper who contains the instance.
		 * @param instance An instance contained into the wrapper.
		 * @param x Coordinate-X of the {@link instance} placement in the {@link wrapper}.
		 * @param y Coordinate-Y of the {@link instance} placement in the {@link wrapper}.
		 * @param z Coordinate-Z of the {@link instance} placement in the {@link wrapper}.
		 */
		public constructor
			(
				wrapper: Wrapper, instance: Instance,
				x: number, y: number, z: number
			);

		/**
		 * Construct from a Wrapper and Instance with its position and orientation. 
		 *
		 * @param wrapper A wrapper who contains the instance.
		 * @param instance An instance contained into the wrapper.
		 * @param x Coordinate-X of the {@link instance} placement in the {@link wrapper}.
		 * @param y Coordinate-Y of the {@link instance} placement in the {@link wrapper}.
		 * @param z Coordinate-Z of the {@link instance} placement in the {@link wrapper}.
		 * @param orientation Placement orientation of wrapped {@link instance}.
		 */
		public constructor
			(
				wrapper: Wrapper, instance: Instance, 
				x: number, y: number, z: number, orientation: number
			);

		public constructor
			(
				wrapper: Wrapper, instance: Instance = null,
				x: number = 0, y: number = 0, z: number = 0, orientation: number = 1
			)
		{
			super();
			
			this.wrapper = wrapper;
			this.instance = instance;
			
			this.x = x;
			this.y = y;
			this.z = z;
			this.orientation = orientation;
		}

		/**
		 * @inheritdoc
		 */
		public construct(xml: library.XML): void
		{
			super.construct(xml);

			let instanceXML = xml.get("instance").at(0);
			this.instance = this.createInstance(instanceXML.getProperty("type"));
			this.instance.construct(instanceXML);
		}

		/**
		 * Factory method of wrapped Instance. 
		 *
		 * @param type Type of contained Instance's type.
		 */
		protected createInstance(type: string): Instance
		{
			if (type == "product")
				return new Product();
			else if (type == "wrapper")
				return new Wrapper();
			else
				return null;
		}

		/* ===========================================================
			SETTERS
				- MEMBERS
				- ESTIMATERS
		==============================================================
			MEMBERS
		----------------------------------------------------------- */
		/**
		 * Set orientation.
		 * 
		 * @param orientation Orientation code (1 to 6).
		 */
		public setOrientation(orientation: number): void
		{
			this.orientation = orientation;
		}
		
		/**
		 * Set position.
		 * 
		 * @param x Coordinate-X of the instance placement in the wrapper.
		 * @param y Coordinate-Y of the instance placement in the wrapper.
		 * @param z Coordinate-Z of the instance placement in the wrapper.
		 */
		public setPosition(x: number, y: number, z: number): void
		{
			this.x = x;
			this.y = y;
			this.z = z;
		}

		/* -----------------------------------------------------------
			ESTIMATERS
		----------------------------------------------------------- */
		/**
		 * @brief Estimate orientation by given size.
		 *
		 * @param width Width by placement.
		 * @param height Height by placement.
		 * @param length Length by placement.
		 */
		public estimateOrientation(width: number, height: number, length: number): void
		{
			if (this.instance.getWidth() == width && this.instance.getHeight() == height)
				this.orientation = 1;
			else if (this.instance.getWidth() == length && this.instance.getHeight() == height)
				this.orientation = 2;
			else if (this.instance.getWidth() == length && this.instance.getHeight() == width)
				this.orientation = 3;
			else if (this.instance.getWidth() == height && this.instance.getHeight() == width)
				this.orientation = 4;
			else if (this.instance.getWidth() == width && this.instance.getHeight() == length)
				this.orientation = 5;
			else
				this.orientation = 6;
		}

		/**
		 * @brief Orientation change is occured in level of the packer.
		 *
		 * @details orientation Packer's new orientation.
		 */
		public changeWrapperOrientation(orientation: number): void
		{
			if (orientation == 1)
				return;

			// DECLARES
			let x: number;
			let y: number;
			let z: number;
			let width: number;
			let height: number;
			let length: number;

			if (orientation == 2)
			{
				width = this.instance.getLength();
				height = this.instance.getHeight();
				length = this.instance.getWidth();

				x = this.z;
				y = this.y;
				z = this.wrapper.getWidth() - (length + this.x);
			}
			else if (orientation == 3)
			{
				width = this.instance.getLength();
				height = this.instance.getWidth();
				length = this.instance.getHeight();

				x = this.z;
				y = this.wrapper.getWidth() - (height + this.x);
				z = this.wrapper.getHeight() - (length + this.y);
			}
			else if (orientation == 4)
			{
				width = this.instance.getHeight();
				height = this.instance.getWidth();
				length = this.instance.getLength();

				x = this.y;
				y = this.x;
				z = this.z;
			}
			else if (orientation == 5)
			{
				width = this.instance.getWidth();
				height = this.instance.getLength();
				length = this.instance.getHeight();

				x = this.x;
				y = this.wrapper.getLength() - (height + this.z);
				z = this.y;
			}
			else
			{
				width = this.instance.getHeight();
				height = this.instance.getLength();
				length = this.instance.getWidth();

				x = this.y;
				y = this.z;
				z = this.wrapper.getWidth() - (length - this.x);
			}

			this.estimateOrientation(width, height, length);
			this.x = x;
			this.y = y;
			this.z = z;
		}

		/* ===========================================================
			GETTERS
		=========================================================== */
		/**
		 * Get wrapper.
		 */
		public getWrapper(): Wrapper
		{
			return this.wrapper;
		}

		/**
		 * Get instance.
		 */
		public getInstance(): Instance
		{
			return this.instance;
		}

		/**
		 * Get x.
		 */
		public getX(): number
		{
			return this.x;
		}

		/**
		 * Get y.
		 */
		public getY(): number
		{
			return this.y;
		}

		/**
		 * Get z.
		 */
		public getZ(): number
		{
			return this.z;
		}

		/**
		 * Get orientation.
		 */
		public getOrientation(): number
		{
			return this.orientation;
		}

		/**
		 * Get width.
		 */
		public getLayoutWidth(): number
		{
			switch (this.orientation)
			{
			case 1:	case 5:
				return this.instance.getWidth();
			case 3:	case 4:
				return this.instance.getHeight();
			default: // 2, 5
				return this.instance.getLength();
			}
		}

		/**
		 * Get height.
		 */
		public getLayoutHeight(): number
		{
			switch (this.orientation)
			{
			case 1:	case 2:
				return this.instance.getHeight();
			case 4:	case 6:
				return this.instance.getWidth();
			default: // 3, 5
				return this.instance.getLength();
			}
		}

		/**
		 * Get length.
		 */
		public getLength(): number
		{
			switch (this.orientation)
			{
			case 1: case 4:
				return this.instance.getLength();
			case 2:	case 3:
				return this.instance.getWidth();
			default: // 5, 6
				return this.instance.getHeight();
			}
		}

		/**
		 * Get volume.
		 */
		public getVolume(): number
		{
			return this.instance.getVolume();
		}

		public get $instanceName(): string
		{
			return this.instance.getName();
		}
		public get $layoutScale(): string
		{
			return this.getLayoutWidth() + ", " + this.getLayoutHeight() + ", " + this.getLength();
		}
		public get $position(): string
		{
			return this.x + ", " + this.y + ", " + this.z;
		}

		/* -----------------------------------------------------------
			EXPORTERS
		----------------------------------------------------------- */
		/**
		 * @inheritdoc
		 */
		public TAG(): string
		{
			return "wrap";
		}

		/**
		 * @inheritdoc
		 */
		public toXML(): library.XML
		{
			let xml: library.XML = super.toXML();
			xml.push(this.instance.toXML());
			
			xml.setProperty("layoutWidth", this.getLayoutWidth() + "");
			xml.setProperty("layoutHeight", this.getLayoutHeight() + "");
			xml.setProperty("layoutLength", this.getLength() + "");

			return xml;
		}
	}
}

File: cpp/src/bws/packer/Wrap.cpp
#include <bws/packer/Wrap.hpp>

#include <bws/packer/Wrapper.hpp>
#include <bws/packer/Instance.hpp>

using namespace std;
using namespace bws::packer;
using namespace samchon::library;
using namespace samchon::protocol;

/* -----------------------------------------------------------
	CONSTRUCTORS
----------------------------------------------------------- */
Wrap::Wrap(Wrapper *wrapper)
	: super()
{
	this->wrapper = wrapper;
	//this->duplicated = 1;
}

Wrap::Wrap
	(
		Wrapper *wrapper, shared_ptr<Instance> instance,
		double x, double y, double z, int direction
	) : Wrap(wrapper)
{
	this->instance = instance;

	this->x = x;
	this->y = y;
	this->z = z;

	this->orientation = direction;
}

void Wrap::construct(shared_ptr<XML> xml)
{
	// 키값을 통해 WRAPPER와 PRODUCT를 가져올 방법을 특정해야 함
	// Wrapper->Packer->productArray로 가도록 한다

	x = xml->getProperty<double>("x");
	y = xml->getProperty<double>("y");
	z = xml->getProperty<double>("z");

	orientation = xml->getProperty<int>("orientation");
	//duplicated = xml->getProperty<size_t>("duplicated");
}

void Wrap::setPosition(double x, double y, double z)
{
	this->x = x;
	this->y = y;
	this->z = z;
}
void Wrap::setOrientation(int val)
{
	this->orientation = val;
}

void Wrap::estimateOrientation(double width, double height, double length)
{
	if (instance->getWidth() == width && instance->getHeight() == height)
		orientation = 1;
	else if (instance->getWidth() == length && instance->getHeight() == height)
		orientation = 2;
	else if (instance->getWidth() == length && instance->getHeight() == width)
		orientation = 3;
	else if (instance->getWidth() == height && instance->getHeight() == width)
		orientation = 4;
	else if (instance->getWidth() == width && instance->getHeight() == length)
		orientation = 5;
	else
		orientation = 6;
}

void Wrap::changeWrapperOrientation(int orietation)
{
	if (orientation == 1)
		return;

	// DECLARES
	double x, y, z;
	double width, height, length;

	if (orientation == 2)
	{
		width = instance->getLength();
		height = instance->getHeight();
		length = instance->getWidth();

		x = this->z;
		y = this->y;
		z = wrapper->getWidth() - (length + this->x);
	}
	else if (orientation == 3)
	{
		width = instance->getLength();
		height = instance->getWidth();
		length = instance->getHeight();

		x = this->z;
		y = wrapper->getWidth() - (height + this->x);
		z = wrapper->getHeight() - (length + this->y);
	}
	else if (orientation == 4)
	{
		width = instance->getHeight();
		height = instance->getWidth();
		length = instance->getLength();

		x = this->y;
		y = this->x;
		z = this->z;
	}
	else if (orientation == 5)
	{
		width = instance->getWidth();
		height = instance->getLength();
		length = instance->getHeight();

		x = this->x;
		y = wrapper->getLength() - (height + this->z);
		z = this->y;
	}
	else
	{
		width = instance->getHeight();
		height = instance->getLength();
		length = instance->getWidth();

		x = this->y;
		y = this->z;
		z = wrapper->getWidth() - (length - this->x);
	}

	this->estimateOrientation(width, height, length);
	this->x = x;
	this->y = y;
	this->z = z;
}

/* -----------------------------------------------------------
	GETTERS
----------------------------------------------------------- */
auto Wrap::getWrapper() const -> Wrapper*
{
	return wrapper;
}
auto Wrap::getInstance() const -> shared_ptr<Instance>
{
	return instance;
}

auto Wrap::getOrientation() const -> int
{
	return orientation;
}

auto Wrap::getX() const -> double
{
	return x;
}
auto Wrap::getY() const -> double
{
	return y;
}
auto Wrap::getZ() const -> double
{
	return z;
}

auto Wrap::getWidth() const -> double
{
	switch (orientation)
	{
	case 1: case 5:
		return instance->getWidth();
	case 3: case 4:
		return instance->getHeight();
	default: // 2, 5
		return instance->getLength();
	}
}
auto Wrap::getHeight() const -> double
{
	switch (orientation)
	{
	case 1: case 2:
		return instance->getHeight();
	case 4: case 6:
		return instance->getWidth();
	default: // 3, 5
		return instance->getLength();
	}
}
auto Wrap::getLength() const -> double
{
	switch (orientation)
	{
	case 1: case 4:
		return instance->getLength();
	case 2: case 3:
		return instance->getWidth();
	default: // 5, 6
		return instance->getHeight();
	}
}
auto Wrap::getVolume() const -> double
{
	return instance->getVolume();
}

/* -----------------------------------------------------------
	EXPORTERS
----------------------------------------------------------- */
auto Wrap::TAG() const -> string
{
	return "wrap";
}

auto Wrap::toXML() const -> shared_ptr<XML>
{
	shared_ptr<XML> &xml = super::toXML();
	xml->setProperty("instance", instance->key());

	// 나중에 없애야 함
	xml->push_back(instance->toXML());

	xml->setProperty("x", x);
	xml->setProperty("y", y);
	xml->setProperty("z", z);
	xml->setProperty("orientation", orientation);
	//xml->setProperty("duplicated", duplicated);

	return xml;
}